In [ ]:
import torch as t
import torchvision.datasets as datasets 
import torchvision.transforms as transforms
import torch.nn as nn

transform = transforms.Compose([
    transforms.ToTensor(), 
    transforms.Normalize((0.5,), (0.5,)) 
])

train_dataset = datasets.FashionMNIST(
    root='./data',
    train=True,
    download=True,
    transform=transform
)

test_dataset = datasets.FashionMNIST(
    root='./data',
    train=False,
    download=True,
    transform=transform
)

train_loader = t.utils.data.DataLoader(train_dataset, batch_size=64, shuffle=True)
test_loader = t.utils.data.DataLoader(test_dataset, batch_size=64, shuffle=False)

print(f"Training samples: {len(train_dataset)}")
print(f"Testing samples: {len(test_dataset)}")

Training samples: 60000
Testing samples: 10000


### notes :
#### 1- no need to suffle test set 
#### 2- train loder call "t.utils.data.DataLoader" , which have an arg calld "train_dataset" this arg is an instance of "datasets.FashionMNIST" which can call FashionMNIST

#### 3- "transform" , cheak that : data is Tensors and normalized 

In [ ]:
class Net(nn.Module):
    def __init__(self):
        super(Net,self).__init__()
        self.linear1 = nn.Linear(28*28, 256) 
        self.linear2 = nn.Linear(256, 128) 
        self.final = nn.Linear(128, 10)
        self.relu = nn.ReLU()

    def forward(self, img): 
        x = img.view(-1, 28*28)
        x = self.relu(self.linear1(x))
        x = self.relu(self.linear2(x))
        x = self.final(x)
        return x
net = Net()

#### no notes , just easy structure model

In [4]:
cross_el = nn.CrossEntropyLoss()
optimizer = t.optim.Adam(net.parameters(), lr=0.001)
epochs = 10  

for epoch in range(epochs):
    net.train()  
    running_loss = 0.0
    for x, y in train_loader:
        optimizer.zero_grad()
        output = net(x)
        loss = cross_el(output, y)
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
    print(f"Epoch [{epoch+1}/{epochs}], Loss: {running_loss/len(train_loader):.4f}")

Epoch [1/10], Loss: 0.4993
Epoch [2/10], Loss: 0.3689
Epoch [3/10], Loss: 0.3313
Epoch [4/10], Loss: 0.3039
Epoch [5/10], Loss: 0.2856
Epoch [6/10], Loss: 0.2712
Epoch [7/10], Loss: 0.2568
Epoch [8/10], Loss: 0.2422
Epoch [9/10], Loss: 0.2325
Epoch [10/10], Loss: 0.2207


In [5]:
correct = 0
total = 0

net.eval()  
with t.no_grad(): 
    for x, y in test_loader:
        output = net(x)  
        
        _, predicted = t.max(output, 1)
        total += y.size(0)
        correct += (predicted == y).sum().item()

accuracy = round(correct / total, 3)
print(f"Accuracy: {accuracy * 100:.1f}%")

Accuracy: 88.4%


---------------------

3. A short conclusion paragraph summarizing model performance.

#### The neural network achieved an accuracy of 88% on the test data.
#### This shows solid performance in recognizing different clothing categories.